### PR TITLE
Adding copy source URL to context menu for editor, frames, and source

### PR DESCRIPTION
**Pull request by jacobjzhang**
… tree

Associated Issue: #1490

### Summary of Changes

Added the ability to copy source URL to:

* Editor.js
* Frames.js
* SourceTree.js

### Test Plan

Opened to-do MVC app and tested across different options.

- Right clicked editor, copied source.
- Right clicked stack frame, copied source.
- Right clicked left source tree, copied source.

### CODE REVIEW

File: src/components/Editor.js [hunk 1]
```diff
@@ -127,9 +137,13 @@ const Editor = React.createClass({
       click: () => this.props.jumpToMappedLocation(sourceLocation)
     };
 
-    showMenu(event, [
-      jumpLabel
-    ]);
+    let items = [];
```
Comment (darkwing) [inline src/components/Editor.js]: `jumpLabel` can probably be added to the array initially

File: src/components/Frames.js [hunk 2]
```diff
@@ -51,6 +54,27 @@ const Frames = React.createClass({
...
+      disabled: false,
+      click: () => copyToTheClipboard(source.url)
+    };
+
+    let items = [];
+    if (isEnabled("copySource")) {
+      items.push(copySourceUrl);
+    }
+
+    showMenu(event, items);
```
Comment (darkwing) [inline src/components/Frames.js]: What if something like:

```
if (isEnabled("copySource")) {
  showMenu(event, [copySourceUrl]);
}
```

No need for `items`

**Comment by jasonLaster** [comment]
ping @jacobjzhang do you have time to finish this up?

File: src/components/SourcesTree.js [hunk 3]
```diff
@@ -119,6 +121,27 @@ let SourcesTree = React.createClass({
...
+    event.stopPropagation();
+    event.preventDefault();
+
+    const source = item.path;
+    const copySourceUrl = {
+      id: "node-menu-copy-source",
+      label: "Copy Source URL",
+      accesskey: "X",
+      disabled: false,
+      click: () => copyToTheClipboard(source)
```
Comment (jacobjzhang) [inline src/components/SourcesTree.js]: Not sure how to fix this, the formatting of item.path is different from the others but it's probably not safe to just format item.path?

File: src/components/Frames.js [hunk 4]
```diff
@@ -60,6 +84,7 @@ const Frames = React.createClass({
           "selected": selectedFrame && selectedFrame.id === frame.id
         }),
         onMouseDown: () => selectFrame(frame),
```
Comment (jasonLaster) [inline src/components/Frames.js]: hmm, you'll have to be clever here i guess :)

File: src/components/SourcesTree.js (hunk 3 above)
Comment (jasonLaster) [inline src/components/SourcesTree.js]: @arthur801031 any ideas?

I'd start by looking at [utils/source-tree](https://github.com/devtools-html/debugger.html/blob/master/src/utils/sources-tree.js#L99)

Comment (arthur801031) [inline src/components/SourcesTree.js]: This works:

```
const sourceURL = item.contents.get("url");
const copySourceUrl = {
      id: "node-menu-copy-source",
      label: "Copy Source URL",
      accesskey: "X",
      disabled: false,
      click: () => copyToTheClipboard(sourceURL)
};
```

File: src/components/Frames.js [hunk 5]
```diff
@@ -51,6 +54,27 @@ const Frames = React.createClass({
     });
   },
 
+  onContextMenu(event, frame) {
+    event.stopPropagation();
+    event.preventDefault();
```
Comment (jacobjzhang) [inline src/components/Frames.js]: Leaving as I believe this is a new issue.

File: src/components/SourcesTree.js [hunk 6]
```diff
@@ -119,6 +121,27 @@ let SourcesTree = React.createClass({
     return Svg("folder");
   },
 
+  onContextMenu(event, item) {
+    event.stopPropagation();
+    event.preventDefault();
```
Comment (jacobjzhang) [inline src/components/SourcesTree.js]: Completed with isDirectory function.

File: src/components/Editor.js [hunk 7]
```diff
@@ -106,6 +107,15 @@ const Editor = React.createClass({
     event.stopPropagation();
     event.preventDefault();
 
+    const source = this.props.selectedSource;
+    const copySourceUrl = {
```
Comment (jacobjzhang) [inline src/components/Editor.js]: I couldn't get showSource to work, currently commented out.

**Comment by jacobjzhang** [comment]
Pushed up what I had for now as to not delay the project further, need help with the Frames click behavior and getting showSource to work in Editor. Initial issue has been resolved and tested.

File: src/components/Editor.js (hunk 7 above)
Comment (jasonLaster) [inline src/components/Editor.js]: not a big deal

File: src/components/Frames.js [hunk 8]
```diff
@@ -59,14 +83,22 @@ const Frames = React.createClass({
...
+        onContextMenu: (e) => this.onContextMenu(e, frame),
         tabIndex: 0
       },
       renderFrameTitle(frame),
       renderFrameLocation(frame)
     );
   },
 
+  onMouseDown(e) {
+    if (e.which == 3) {
```
Comment (jasonLaster) [inline src/components/Frames.js]: I think we need to check to see if its the selected frame

File: src/components/SourcesTree.js [hunk 9]
```diff
@@ -119,6 +121,31 @@ let SourcesTree = React.createClass({
...
+        click: () => copyToTheClipboard(source)
+      };
+
+      let items = [];
+
+      if (isEnabled("copySource")) {
+        items.push(copySourceUrl);
+      }
+
+      showMenu(event, items);
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: hmm, lets move `let items` and `showMenu(event, items);` out of the the condition

**Comment by jacobjzhang** [comment]
@jasonLaster can I still wrap this up?

**Comment by jasonLaster** [comment]
@jacobjzhang go for it

**Comment by jacobjzhang** [comment]
Squashed the two commits, did a rebase on master, cherry-picked them over. Did some cleanup, some basic testing (source code URL copy looks good on all 3 panels), then worked on ensuring the right-click on a non-selected frame does not switch it over. Should be good now, let me know if anything else needs to be fixed.

**Comment by jasonLaster** [comment]
@jacobjzhang looking great. Found some flow issues, that are mostly minor. 

Try running `yarn run flow` locally and adding some `isRequired` calls https://github.com/devtools-html/debugger.html/blob/master/docs/flow.md#required-property

**Comment by jacobjzhang** [comment]
Cleaned up flow errors and tested. Unfortunately a lot of the errors were related to frames/sources which aren't required (e.g. there can be no frames, hence no frame source), so isRequired wasn't that useful.

File: src/components/Editor/index.js [hunk 10]
```diff
@@ -173,7 +183,20 @@ const Editor = React.createClass({
...
+    }
+
+    const showSourceMenuItem = {
+      id: "node-menu-show-source",
+      label: "Show source",
+      accesskey: "s",
+      disabled: false,
+      click: () => showSource(tab)
+    };
+
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: Hmm, it looks like `tab` and `showSource` are not defined in this function

File: src/components/SecondaryPanes/Frames.js [hunk 11]
```diff
@@ -72,14 +100,24 @@ const Frames = React.createClass({
...
+        onMouseDown: (e) => this.onMouseDown(e, frame, selectedFrame),
+        onContextMenu: (e) => this.onContextMenu(e, frame),
         tabIndex: 0
       },
       renderFrameTitle(frame),
       renderFrameLocation(frame)
     );
   },
 
+  onMouseDown(e, frame, selectedFrame) {
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/Frames.js]: could you add types here. See `EditorSearchBar` for event handler types.

File: src/components/SecondaryPanes/Frames.js [hunk 12]
```diff
@@ -72,14 +100,24 @@ const Frames = React.createClass({
...
+    if (e.nativeEvent.which == 3 
+        && selectedFrame
+        && selectedFrame.id != frame.id) {
+      return;
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/Frames.js]: do we need `&& selectedFrame` if we know it's required.

Comment (jacobjzhang) [inline src/components/SecondaryPanes/Frames.js]: this was my solution to flowtype complaining about selectedFrame being possibly undefined (which is valid, user doesn't need to select a frame)

**Comment by jasonLaster** [comment]
Hmm. Try typing the params.

Therell always be a frame

File: src/components/SecondaryPanes/Frames.js (hunk 12 above)
Comment (jacobjzhang) [inline src/components/SecondaryPanes/Frames.js]: Ah. I stand corrected. Put an isRequired on selectedFrame

File: src/components/Editor/index.js (hunk 10 above)
Comment (jacobjzhang) [inline src/components/Editor/index.js]: I just realized, does it make sense to have showSource in the Editor? We'd already be on the correct document when the context menu comes up (assuming showSource literally just shows the source of the file)

**Comment by jacobjzhang** [comment]
Cleaned up flowtypes and did some extra testing. As discussed, may need to open separate issue for showSource inconsistency.

File: src/components/Editor/index.js [hunk 13]
```diff
@@ -122,17 +123,29 @@ const Editor = React.createClass({
...
     event.stopPropagation();
     event.preventDefault();
 
     const isMapped = await hasMappedSource(selectedLocation);
 
+    const source = this.props.selectedSource;
+    const copySourceUrl = {
+      id: "node-menu-copy-source",
+      label: "Copy Source URL",
+      accesskey: "X",
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: @jacobjzhang the label and accessKey need L10N as well

**Comment by jacobjzhang** [comment]
Added copySourceUrl to L10N in all use cases. Tested.

**Pull request merged**